### PR TITLE
feat: Add GradMax initialization method for neuron addition

### PR DESCRIPTION
**Pull request by stephane-rivaud**
## Summary

This PR adds GradMax-compatible neuron-growth behavior to the growth pipeline and unifies growth controls under explicit primitive options.

Instead of a string-based method selector, the public API now exposes primitive flags on `compute_optimal_updates(...)` that can express both TINY-style and GradMax-style behavior.

## Background

GradMax (Evci et al., 2022, arXiv:2201.05125) initializes new neurons to maximize gradient norm. In this PR, that behavior is represented through explicit option combinations rather than a dedicated `initialization_method` argument.

### Breaking Changes

1. Public `compute_optimal_added_parameters()` calls are replaced by `compute_optimal_updates()`.
2. `compute_optimal_updates()` is now the public entry point and returns `(alpha_weight, alpha_bias)`.
3. Internal/private `_compute_optimal_added_parameters()` and related helpers use primitive options:
   - `compute_delta`
   - `use_covariance`
   - `alpha_zero`
   - `omega_zero`
   - `use_projection`
   - `ignore_singular_values`

### Migration Guide

**Old call pattern (removed public API):**

```python
alpha_weight, alpha_bias, omega, eigenvalues = layer.compute_optimal_added_parameters(...)
```

**New call pattern:**

```python
alpha_weight, alpha_bias = layer.compute_optimal_updates(
    numerical_threshold=1e-6,
    statistical_threshold=1e-3,
    maximum_added_neurons=5,
    compute_delta=True,
    use_covariance=True,
    alpha_zero=False,
    omega_zero=False,
    use_projection=True,
    ignore_singular_values=False,
)
```

Common presets with primitive flags:

- TINY-style: `compute_delta=True`, `use_covariance=True`, `alpha_zero=False`, `omega_zero=False`, `use_projection=True`, `ignore_singular_values=False`
- GradMax-style: `compute_delta=False`, `use_covariance=False`, `alpha_zero=True`, `omega_zero=False`, `use_projection=False`, `ignore_singular_values=True`

If you need `omega` or singular values after calling `compute_optimal_updates(...)`, use module attributes populated during the update flow (for example `extended_input_layer` and `eigenvalues_extension`).

## Implementation Notes

- `src/gromo/utils/tools.py`: keeps the shared SVD-based alpha/omega computation with explicit guardrails for invalid option combinations.
- `src/gromo/modules/growing_module.py`: exposes `compute_optimal_updates(...)` as the public orchestration API and routes into private helpers.
- `src/gromo/containers/*`: migrated to the new public update API.

## Testing

- Updated and extended unit/integration tests for:
  - primitive-option dispatch and validation
  - side-effect behavior (`parameter_update_decrease`, `first_order_improvement`)
  - no-previous-module edge paths
  - conv/pooling integration tolerance rationale under unified defaults

## Documentation

- Updated `docs/source/whats_new.rst` to reflect the actual API shape and migration path in this PR.

## References

- Evci et al. (2022), "Gradmax: Growing neural networks using gradient information", arXiv:2201.05125
- Rudkiewicz et al. (2024), "Growing Tiny Networks", arXiv:2405.19816

### CODE REVIEW

**Comment by TheoRudkiewicz** [comment]
I am not sure of what is the goal of the examples. If the goal is to integrate it in the documentation like [https://growingnet.github.io/gromo/auto_examples/plot_growing_container_tutorial.html](https://growingnet.github.io/gromo/auto_examples/plot_growing_container_tutorial.html)  they should run very fast and be formatted like the other ones (ie in rst style and named with the correct filename pattern  `(/plot_|/tutorial_)*`. I believe this should not necessary be the goal. It could be kept as example for user that download the code. In this case it would be good to make this distinction clear and maybe put them in a sub folder?

[Out of this PR scope] Moreover examples are a bit a mess and only the `*_tutorial.py` files are a bit clean.  And we should clean the examples and the `auxilliary_functions` file. Moreover to avoid multiple version of train, evalutate, extended evaluate, line search... we should start to consider moving some of those functions from experimental grow to gromo and uses those in the examples...

File: src/gromo/modules/growing_module.py [hunk 1]
```diff
@@ -1865,6 +1898,7 @@ def compute_optimal_updates(
                 update_previous=update_previous,
                 dtype=dtype,
                 use_projected_gradient=use_projected_gradient,
+                initialization_method=initialization_method,
             )
```
Comment (TheoRudkiewicz) [inline src/gromo/modules/growing_module.py]: Missing `initialization_method` parameter in `compute_optimal_added_parameters`.

File: src/gromo/containers/growing_block.py [hunk 2]
```diff
@@ -344,15 +344,19 @@ def set_scaling_factor(self, factor: float) -> None:
 
     def compute_optimal_updates(
         self,
-        numerical_threshold: float = 1e-15,
+        numerical_threshold: float = 1e-6,
         statistical_threshold: float = 1e-3,
         maximum_added_neurons: int | None = None,
+        update_previous: bool = True,
         dtype: torch.dtype = torch.float32,
```
Comment (TheoRudkiewicz) [inline src/gromo/containers/growing_block.py]: I don't think we want this option.

File: src/gromo/containers/growing_block.py [hunk 3]
```diff
@@ -392,24 +396,96 @@ def compute_optimal_updates(
...
+            if config["use_projection"]:
+                warn(
+                    f"GrowingBlock with hidden_neurons=0 cannot use projection "
+                    f"(tensor_n requires delta_raw from compute_optimal_delta()). "
+                    f"Method '{initialization_method}' expects use_projection=True, "
+                    f"but will use use_projection=False (raw gradient -tensor_m_prev). "
+                    f"This may result in different behavior than expected.",
+                    UserWarning,
+                    stacklevel=2,
+                )
```
[the remark this answers is not in the document]
Comment (TheoRudkiewicz) [reply src/gromo/containers/growing_block.py]: Very good. I slighlthly changed the comment.
```suggestion
            # With hidden_neurons == 0 we cannot compute tensor_n (delta_raw from  
            # compute_optimal_delta() is unavailable). We force use_projection=False  
            # here as the bottlenck is the gradient in this case (null-space manifold).  
```

File: src/gromo/modules/conv2d_growing_module.py [hunk 4]
```diff
@@ -1682,8 +1688,14 @@ def compute_optimal_added_parameters(
...
-        use_projected_gradient: bool
-            whereas to use the projected gradient ie `tensor_n` or the raw `tensor_m`
+        use_covariance: bool
+            if True, use S matrix (covariance preconditioning), else use Identity
+        alpha_zero: bool
+            if True, set alpha (incoming weights) to zero, else compute from SVD
+        use_projection: bool
+            if True, use projected gradient (tensor_n), else use raw gradient (-tensor_m_prev)
+        **kwargs : Any
+            Additional keyword arguments passed to the parent method.
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/conv2d_growing_module.py]: Addressed in commit 933b2e2.

I aligned nested-call kwargs behavior and docs:
- `LinearGrowingModule._compute_optimal_added_parameters` now accepts `**kwargs` for compatibility with `GrowingModule.compute_optimal_updates(..., **kwargs)` passthrough.
- Conv2d and related docstrings now explicitly describe kwargs as nested propagation for specific sub-function calls.

Also added regression coverage in `tests/test_linear_growing_module.py` for extra kwargs passthrough.

File: src/gromo/utils/tools.py [hunk 5]
```diff
@@ -218,6 +220,82 @@ def compute_optimal_added_parameters(
...
+    # SVD(M^T) = U Σ V^T where U is (out_features, min(out_features, in_features))
+    try:
+        u, s, _ = torch.linalg.svd(tensor_m_prev.T, full_matrices=False)
+    except torch.linalg.LinAlgError:
+        print("Warning: An error occurred during the SVD computation in GradMax.")
+        print(
+            f"tensor_m_prev: {tensor_m_prev.min()=}, "
+            f"{tensor_m_prev.max()=}, {tensor_m_prev.shape=}"
+        )
+        u, s, _ = torch.linalg.svd(tensor_m_prev.T, full_matrices=False)
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/utils/tools.py]: Resolved: the SVD retry anti-pattern is no longer present. We keep diagnostics and re-raise without re-running the same failing operation.

File: src/gromo/modules/growing_module.py [hunk 6]
```diff
@@ -2139,54 +2166,79 @@ def first_order_improvement(self) -> torch.Tensor:
 
     def compute_optimal_updates(
         self,
-        numerical_threshold: float = 1e-10,
-        statistical_threshold: float = 1e-5,
+        numerical_threshold: float = 1e-6,
+        statistical_threshold: float = 1e-3,
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/growing_module.py]: Addressed in commit 8397730. I removed the backward-compatibility wording and documented this as an intentional default unification: numerical_threshold=1e-6 and statistical_threshold=1e-3 across growth APIs, with rationale in whats_new.rst.

File: tests/test_growing_module.py [hunk 7]
```diff
@@ -777,115 +813,236 @@ def test_isinstance_merge_growing_module_check(self):
...
-        loss = torch.norm(output2)
+        # Forward/backward pass to gather statistics
+        x = torch.randn(2, 3, device=global_device())
+        y = first_layer(x)
+        y = second_layer(y)
+        loss = torch.norm(y)
         loss.backward()
+        first_layer.update_computation()
+        second_layer.update_computation()
+
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply tests/test_growing_module.py]: Fixed locally in `tests/test_growing_module.py`.
I replaced the stale no-op test with `test_compute_optimal_updates_rejects_legacy_method_kwarg`, which asserts:
`compute_optimal_updates(initialization_method="tiny")` raises `TypeError` (unexpected keyword argument).
This makes the test assertion-bearing and aligned with the current primitive-options API (no `initialization_method` parameter).

File: src/gromo/modules/growing_module.py [hunk 8]
```diff
@@ -2139,54 +2180,125 @@ def first_order_improvement(self) -> torch.Tensor:
...
+        Compute the optimal update and additional neurons.
+
+        This method computes optimal weight updates for growing neural networks
+        by analyzing gradient statistics and covariance information.
+
+        Hyper-parameters to reproduce papers:
+        -------------------------------------
+
+        - TINY (Growing Tiny Networks: Spotting Expressivity Bottlenecks and Fixing Them Optimally)
+        compute_delta: bool = False,
```
[the remark this answers is not in the document]
Comment (stephane-rivaud) [reply src/gromo/modules/growing_module.py]: The experiments in the TINY paper is equivalent to `compute_delta=False`.
